src/number_theory/bernoulli.lean

```diff
@@ -258,3 +267,80 @@ begin
   { ext x, rw bernoulli_eq_bernoulli' (succ_ne_zero x ∘ succ.inj) },
   { ring },
 end
+
+lemma bernoulli_spec' (n: ℕ) :
+  ∑ k in finset.nat.antidiagonal n,
+  ((k.1 + k.2).choose k.2 : ℚ) / (k.2 + 1) * bernoulli k.1 = if n = 0 then 1 else 0 :=
+begin
+  cases n,
+  { simp },
+  rw if_neg (succ_ne_zero _),
+  -- algebra facts
+  have h₁ : (1, n) ∈ nat.antidiagonal n.succ := by simp [nat.mem_antidiagonal, add_comm],
+  have h₂ : (n:ℚ) + 1 ≠ 0 := by exact_mod_cast succ_ne_zero _,
+  have h₃ : (1 + n).choose n = n + 1 := by simp [add_comm],
+  -- key equation:  the corresponding fact for `bernoulli'`
+  have H := bernoulli'_spec' n.succ,
+  -- massage it to match the structure of the goal, then convert piece by piece
+  rw ← finset.add_sum_diff_singleton h₁ at H ⊢,
+  apply add_eq_of_eq_sub',
+  convert eq_sub_of_add_eq' H using 1,
+  { apply sum_congr rfl,
+    intros p h,
+    obtain ⟨h', h''⟩ : p ∈ _ ∧ p ≠ _ := by rwa [mem_sdiff, mem_singleton] at h,
+    have : p.fst ≠ (1, n).fst := by simpa [h''] using nat.antidiagonal_congr h' h₁,
+    simp [this, bernoulli_eq_bernoulli'] },
+  { field_simp [h₃],
+    norm_num }
+end
+
+theorem bernoulli_power_series :
+  power_series.mk (λ n, algebra_map ℚ A (bernoulli n / n!)) * (exp A - 1) = X :=
```

Comment:
Should the factor on the left be made into a `def`? Something like (sketch)
```
def bernoulli_power_series :=
power_series.mk (λ n, algebra_map ℚ A (bernoulli n / n!))

theorem bernoulli_power_series_mul_exp_sub_one :
  bernoulli_power_series * (exp A - 1) = X :=
```

Another option is to have some extra constructors for generating functions. Because the `f n / n!` pattern is quite common. So maybe instead of the above def, we could have
```
def ordinary_generating_function (f) := power_series.mk  (λ n, algebra_map ℚ A (f n))

def exponential_generating_function (f) := power_series.mk  (λ n, algebra_map ℚ A (f n / n!))
```

Reply:
Re making the factor on the left as a separate def: good idea, I'll do that.

About the other suggestion: I think it would make perfect sense to have ogf and egf available as definitions. However this I guess this would require more changes in many more places, e.g. one could then define `exp A` as an `exponential_generating_function`. I think it might be better to do that in a separate pull request.

Reply:
> ```
> def bernoulli_power_series :=
> power_series.mk (λ n, algebra_map ℚ A (bernoulli n / n!))
> 
> theorem bernoulli_power_series_mul_exp_sub_one :
>   bernoulli_power_series * (exp A - 1) = X :=
> ```
> 
Tried to add something along those line, but had trouble proving then using that definition `bernoulli'_power_series` in a proof. How can you simply say: replace lhs of def with rhs in an expression you are trying to prove.
I made it work by introducing a lemma `bernoulli'_power_series_def`, but this is probably not the way to go.

Reply:
You can `rw [name_of_the_def]` usually, but your solution is also fine. But maybe what's really need is a lemma `coeff_bernoulli'_power_series`?

Reply:
Now I use an additional `suffices` in the proof of `bernoulli'_odd_eq_zero` and use `dsimp` to replace the new definition of `bernoulli'_power_series`.